### ISSUE TITLE
PunchAngle not being properly predicted/networked

### ISSUE THREAD

**Issue by Kefta** [Facepunch/garrysmod-issues#2887]
Tested on a dedicated server. Notice that command 726401 predicts again AFTER 726402 starts, and the punch value changes between these cases. Also, the values never match with the server even though both were set to the same initial punch value.

CLIENT
```
726400
IsFirstTimePredicted: true
Punch: 0.458 0.132 0.000

726400
IsFirstTimePredicted: false
Punch: 0.458 0.132 0.000

726401
IsFirstTimePredicted: true
Punch: 0.553 0.162 0.000

726401
IsFirstTimePredicted: false
Punch: 0.553 0.162 0.000

726402
IsFirstTimePredicted: true
Punch: 0.627 0.185 0.000

726401
IsFirstTimePredicted: false
Punch: 0.563 0.156 0.000

726402
IsFirstTimePredicted: false
Punch: 0.636 0.179 0.000

726402
IsFirstTimePredicted: false
Punch: 0.636 0.179 0.000
```

SERVER
```
726400
IsFirstTimePredicted: true
Punch: 0.475 0.138 0.000

726401
IsFirstTimePredicted: true
Punch: 0.569 0.168 0.000

726402
IsFirstTimePredicted: true
Punch: 0.643 0.191 0.000
```

**Comment by Faerachnidendro** [comment]
Well I'll be damned... that explains the jitter on weapon recoil a bit more, for some weapon bases.
Good spotting.

**Comment by Jvs34** [comment]
No, the jitter on weapon recoil of the weapon bases you try is because those weapons don't predict properly.

The prediction of the punch angle has a margin of error that it accepts so it will never jitter unless the values are much bigger than the margin of error and it will restore back to the value of the server.

**Comment by Faerachnidendro** [comment]
Good to know!
Mine in particular are running far smoother anyway, thanks to util.SharedRandom.

**Comment by Kefta** [comment]
I believe the inaccuracies come from the tolerance mixed with network float inaccuracies, which essentially lead to it becoming more worse between realms. See here:

![punch](https://cloud.githubusercontent.com/assets/4593679/20482270/600f13e8-afb1-11e6-871b-1647abb8c375.png)

Although both start out the same, they quickly become out of sync with the DTVar in the Move hook storing the angle becomes further away between realms.

**Comment by Faerachnidendro** [comment]
@Kefta was this using util.SharedRandom? would be interesting to see a comparison - might do it later myself.

**Comment by Kefta** [comment]
No, I use my own shared minstd_rand library, but it played no part in this demonstration. An initial shared value for the punch was set and it slowly became more out of sync. I'll post my decay code later today

**Comment by Faerachnidendro** [comment]
o_o
Now I'm curious about your library! 😝 
If you don't mind, I'll end up using your decay code on my end, to see if there's any difference with the variation between different machines(thinking that perhaps hardware(including network) capabilities might have an affect on the outcome).

**Comment by willox** [comment]
I'm struggling to make a weapon that can reproduce this

**Comment by Kefta** [comment]
### Default decay
Test code: http://pastebin.com/wAAFiyrV
Default SMG (out of sync by 18572): http://pastebin.com/WBipWTQm
weapon_base inherent (http://pastebin.com/xUcYGaHt): http://pastebin.com/RApxKDRq

### Custom decay
Decay code: http://pastebin.com/HcBm5MFR
weapon_base inherent: http://pastebin.com/taDpzUSf

These are obviously very small differences that don't matter too much alone, but spread algorithms that place a higher priority on punch angles (CS:S for example places twice the normal priority) and firing before a viewpunch angle fully decays creates even larger inaccuracies.

This is the output with prediction errors and impacts enabled of my weapon base with the custom decaying enabled (https://github.com/Kefta/GSWeapons/blob/master/gsweapons_css/lua/weapons/weapon_csbase_rifle.lua#L120-L169) which sets the viewpunch to the same initial value: http://pastebin.com/4KDcVH29
![impact](https://cloud.githubusercontent.com/assets/4593679/20552336/82b7e38e-b10f-11e6-98e9-c46846fa9ebc.png)

**Comment by Kefta** [comment]
Probably unrelated, but I've noticed that the punch angle caused by fall damage will desync on the pitch, but the roll is within acceptable error range.

002 CPlayerLocalData::m_vecPunchAngleVel - vec[] differs (1st diff) (net 4.812500 0.000000 -16.687500 - pred 0.000000 0.000000 -16.706417) delta(-4.812500 0.000000 -0.018917)
001 CPlayerLocalData::m_vecPunchAngle - vec[] differs (1st diff) (net -2.000000 0.000000 8.687500 - pred 0.000000 0.000000 8.716500) delta(2.000000 0.000000 0.029000)

**Comment by wrefgtzweve** [comment]
Seems to be still a issue, even with basegame hl2 weapons like the ar2, smg and 357:
![Image](https://github.com/user-attachments/assets/6b5f435a-4155-4bb7-97c3-aaa098c74863)

```lua
local SERVER = SERVER
local CLIENT = CLIENT

local boxColor = SERVER and Color( 255, 0, 0, 50 ) or Color( 0, 255, 0, 50 )
local bulletColor = Color( 0, 0, 255, 100 )
local missColor = Color( 255, 0, 0, 100 )
hook.Add( "PostEntityFireBullets", "ShowBulletTracers", function( ent, data )
    if not ent:IsPlayer() then return end
    
    local tr = data.Trace
    debugoverlay.Line( tr.StartPos, tr.HitPos, 30, boxColor, true )
    debugoverlay.Box( tr.HitPos, Vector( -1, -1, -1 ), Vector( 1, 1, 1 ), 30, boxColor, true )
end )
```

I was having this issue with my own weapons when i was doing:
```lua
local bulletDir = ( owner:GetAimVector():Angle() + owner:GetViewPunchAngles() ):Forward()
```

Synced perfectly when i changed it to:
```lua
local bulletDir = ( owner:GetAimVector():Angle() ):Forward()
```

**Comment by Grocel** [comment]
Reporting this here from #6488.

<img width="1697" height="329" alt="Image" src="https://github.com/user-attachments/assets/b3a2f985-dec7-4f39-b8bb-19c30fc23484" />

**Comment by robotboy655** [comment]
Should be fixed on `network_test` beta.

**Comment by wrefgtzweve** [comment]
> Should be fixed on `network_test` beta.

Can confirm, works 😄 

<img width="1368" height="1004" alt="Image" src="https://github.com/user-attachments/assets/301439fb-e122-4e7b-bcdf-4d9e03115afb" />
